include/svs/quantization/scalar/scalar.h

```diff
@@ -0,0 +1,475 @@
+    SQDataset(size_t size, size_t dims)
+        : data_{size, dims} {}
+    SQDataset(data_type data, float scale, float bias)
+        : scale_(scale)
+        , bias_(bias)
+        , data_{std::move(data)} {}
+
+    size_t size() const { return data_.size(); }
+    size_t dimensions() const { return data_.dimensions(); }
+
+    float get_scale() const { return scale_; }
+    float get_bias() const { return bias_; }
+
+    const_value_type get_datum(size_t i) const { return data_.get_datum(i); }
+
+    std::vector<float> decompress_datum(size_t i) const {
+        auto datum = get_datum(i);
+        std::vector<float> buffer(datum.size());
+        std::transform(datum.begin(), datum.end(), buffer.begin(), [&](element_type v) {
+            return detail::decompress(v, scale_, bias_);
+        });
+        return buffer;
+    }
+
+    template <typename QueryType, size_t N>
+    void set_datum(size_t i, std::span<QueryType, N> datum) {
+        auto dims = dimensions();
+        assert(datum.size() == dims);
+
+        // Compress elements
+        std::vector<element_type> buffer(dims);
+        std::transform(datum.begin(), datum.end(), buffer.begin(), [&](QueryType v) {
+            return detail::compress<QueryType, element_type>(v, scale_, bias_);
+        });
+
+        data_.set_datum(i, buffer);
+        // TODO: Float16 truncation check? (see codec.h, line 1[14)
+    }
+
+    template <data::ImmutableMemoryDataset Dataset>
+    static SQDataset compress(const Dataset& data, const allocator_type& allocator = {}) {
+        return compress(data, 1, allocator);
+    }
+
+    template <data::ImmutableMemoryDataset Dataset>
+    static SQDataset compress(
+        const Dataset& data, size_t num_threads, const allocator_type& allocator = {}
+    ) {
+        auto pool = threads::DefaultThreadPool{num_threads};
+        return compress(data, pool, allocator);
+    }
+
+    template <data::ImmutableMemoryDataset Dataset, threads::ThreadPool Pool>
+    static SQDataset
+    compress(const Dataset& data, Pool& threadpool, const allocator_type& allocator = {}) {
+        if (Extent != Dynamic && data.dimensions() != Extent) {
+            throw ANNEXCEPTION("Dimension mismatch!");
+        }
+
+        // Get dataset extrema
+        auto minmax = detail::MinMax{};
+        auto global = minmax(data, threadpool);
+
+        // Compute scale and bias
+        constexpr float MIN = std::numeric_limits<element_type>::min();
+        constexpr float MAX = std::numeric_limits<element_type>::max();
+        float scale = (global.max - global.min) / (MAX - MIN);
+        float bias = global.min - MIN * scale;
+
+        // Compress data
+        auto compressor = detail::Compressor<element_type, data_type>{scale, bias};
+        auto compressed = compressor(data, threadpool, allocator);
+
+        return SQDataset<element_type, extent, allocator_type>{
+            std::move(compressed), scale, bias};
+    }
+
+    /// @brief Compact the dataset
```

Comment:
Just to keep in mind we need to dd the Doxygen docstrings for the other functions and some more documentation for SQDataset in next PR